iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/iothub/twin/TwinPnPTests.java

```diff
@@ -0,0 +1,185 @@
+package tests.integration.com.microsoft.azure.sdk.iot.iothub.twin;
+
+import com.microsoft.azure.sdk.iot.device.*;
+import com.microsoft.azure.sdk.iot.device.exceptions.ModuleClientException;
+import com.microsoft.azure.sdk.iot.service.RegistryManager;
+import com.microsoft.azure.sdk.iot.service.devicetwin.DeviceTwin;
+import com.microsoft.azure.sdk.iot.service.devicetwin.DeviceTwinDevice;
+import com.microsoft.azure.sdk.iot.service.exceptions.IotHubException;
+import com.microsoft.azure.sdk.iot.service.Device;
+import com.microsoft.azure.sdk.iot.service.Module;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.DeviceConnectionString;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.IntegrationTest;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.TestConstants;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.Tools;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.annotations.IotHubTest;
+import tests.integration.com.microsoft.azure.sdk.iot.helpers.annotations.StandardTierHubOnlyTest;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.*;
+
+import static com.microsoft.azure.sdk.iot.device.IotHubClientProtocol.*;
+import static com.microsoft.azure.sdk.iot.device.IotHubClientProtocol.MQTT_WS;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Test class containing all tests to be run for PnP.
+ */
+@IotHubTest
+@RunWith(Parameterized.class)
+public class TwinPnPTests extends IntegrationTest
+{
+    protected static String iotHubConnectionString = "";
+    private static RegistryManager registryManager;
+    private static final String ModelId = "dtmi:com:test:e2e;1";
+
+    @Parameterized.Parameters(name = "{0}")
+    public static Collection inputs() throws IOException
+    {
+        iotHubConnectionString = Tools.retrieveEnvironmentVariableValue(TestConstants.IOT_HUB_CONNECTION_STRING_ENV_VAR_NAME);
+        isBasicTierHub = Boolean.parseBoolean(Tools.retrieveEnvironmentVariableValue(TestConstants.IS_BASIC_TIER_HUB_ENV_VAR_NAME));
+        isPullRequest = Boolean.parseBoolean(Tools.retrieveEnvironmentVariableValue(TestConstants.IS_PULL_REQUEST));
+
+        registryManager = RegistryManager.createFromConnectionString(iotHubConnectionString);
+
+        List inputs =  Arrays.asList(
+                new Object[][]
+                        {
+                                {MQTT},
+                                {MQTT_WS},
+                        }
```

Comment:
Seems like you could also parameterize Device vs Module here

Reply:
See here for an example: https://github.com/Azure/azure-iot-sdk-java/blob/fc2ed8189b341e4df47d49a07269067a88a123ab/iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/iothub/setup/SendMessagesCommon.java#L63

Reply:
cool, let me check that.

Reply:
Also x509 vs SAS authentication. Since those two have pretty different connect flows, it would be worth testing that

Reply:
done